nexus/src/db/datastore/update.rs

```diff
@@ -56,4 +64,168 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
             .internal_context("deleting outdated available artifacts")
     }
+
+    pub async fn system_update_create(
+        &self,
+        opctx: &OpContext,
+        update: SystemUpdate,
+    ) -> CreateResult<SystemUpdate> {
+        // TODO: what's the right permission here?
+        opctx.authorize(authz::Action::CreateChild, &authz::FLEET).await?;
+
+        use db::schema::system_update::dsl::*;
+
+        diesel::insert_into(system_update)
+            .values(update.clone())
+            .on_conflict(version)
+            .do_nothing()
+            .returning(SystemUpdate::as_returning())
+            .get_result_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::Conflict(
+                        ResourceType::SystemUpdate,
+                        &update.version.to_string(),
+                    ),
+                )
+            })
+    }
+
+    pub async fn component_update_create(
+        &self,
+        opctx: &OpContext,
+        system_update_id: Uuid,
+        update: ComponentUpdate,
+    ) -> CreateResult<ComponentUpdate> {
+        // TODO: what's the right permission here?
+        opctx.authorize(authz::Action::CreateChild, &authz::FLEET).await?;
+
+        use db::schema::component_update;
+        use db::schema::system_update_component_update as join_table;
+
+        let result = diesel::insert_into(component_update::table)
+            .values(update.clone())
+            .on_conflict(component_update::columns::id) // TODO: should also conflict on version
+            .do_nothing()
+            .returning(ComponentUpdate::as_returning())
+            .get_result_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::Conflict(
+                        ResourceType::ComponentUpdate,
+                        // TODO: string representing both system and component updates?
+                        &update.version.to_string(),
+                    ),
+                )
+            });
+
+        diesel::insert_into(join_table::table)
+            .values(SystemUpdateComponentUpdate {
+                system_update_id,
+                component_update_id: update.id(),
+            })
+            .on_conflict(join_table::all_columns)
+            .do_nothing()
+            .returning(SystemUpdateComponentUpdate::as_returning())
```

Comment:
Same as the other case of `ON CONFLICT DO NOTHING RETURNING` . There appear to be a few more of these down below too.